src/flag_gems/ops/scatter.py

```diff
@@ -1,216 +1,300 @@
+
+            for i in range(rank):
+                function_ns.create_name(f"inp_stride_{i}")
+            stride_args = ", ".join(f"inp_stride_{i}: int" for i in range(rank))
+            code.writeline(f"{stride_args}, # stride for inp")
+
+            for i in range(rank):
+                function_ns.create_name(f"index_stride_{i}")
+            stride_args = ", ".join(f"index_stride_{i}: int" for i in range(rank))
+            code.writeline(f"{stride_args}, # stride for index")
+
+            for i in range(rank):
+                function_ns.create_name(f"index_shape_{i}")
+            shape_args = ", ".join(f"index_shape_{i}: int" for i in range(rank))
+            code.writeline(f"{shape_args}, # shape for index")
+
+            code.writeline("dim,")
+            code.writeline("stride_dim,")
+            code.writeline("M,")
+            code.writeline("N,")
+            # reduce options
+            code.writeline("IS_ADD: tl.constexpr,")
+            code.writeline("IS_MUL: tl.constexpr,")
+            code.writeline("BLOCK_M: tl.constexpr,")
+            code.writeline("BLOCK_N: tl.constexpr,")
+
+    code.writeline("):")
+
+    # Kernel Code
+    with code.indent():
+        code.writeline("pid_x = tl.program_id(0)")
+        code.writeline("pid_y = tl.program_id(1)")
+        code.writeline(
+            "rows_offsets = pid_x * BLOCK_M + tl.arange(0, BLOCK_M)[:, None]"
+        )
+        code.writeline(
+            "cols_offsets = pid_y * BLOCK_N + tl.arange(0, BLOCK_N)[None, :]"
+        )
+        code.writeline("rows_mask = rows_offsets < M")
+        code.writeline("cols_mask = cols_offsets < N")
+
+        code.writeline("offsets = rows_offsets * N + cols_offsets")
+        code.writeline("mask = rows_mask & cols_mask")
+
+        #   1. Calculate inp_offsets and idx_offsets
+        code.writeline("inp_offsets = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.int32)")
+        code.writeline("idx_offsets = tl.zeros((BLOCK_M, BLOCK_N), dtype=tl.int32)")
+        code.writeline("cur_idx = rows_offsets * N + cols_offsets")
+
+        #   2. snippets
+        for i in range(rank):
+            code.writeline(f"mod = cur_idx % index_shape_{i}")
+            code.writeline(f"inp_offsets += mod * inp_stride_{i}")
+            code.writeline(f"idx_offsets += mod * index_stride_{i}")
+            if i != (rank - 1):
+                code.writeline(f"cur_idx = cur_idx // index_shape_{i}")
+
+        #   3. Use offsets to scatter
+        code.writeline(
+            "cur_src = tl.load(src_strided + idx_offsets, mask=mask, other=0)"
+        )
+        code.writeline("cur_index = tl.load(index + idx_offsets, mask=mask, other=0)")
+        code.writeline("inp_offsets += cur_index * stride_dim")
+
+        code.newline()
+        code.writeline("if IS_ADD: ")
+        with code.indent():
+            code.writeline(
+                "cur_inp = tl.load(inp + inp_offsets, mask=mask, other=0).to(tl.float32)"
```

Comment:
Why do we need to cast the payload to float32? Should we keep the dtype as what it is?

Reply:
As `scatter` performs arithmetic calculations with the reduction option turning on, upcasting the operands may be necessary to ensure accuracy. 

However, my tests with the original dtype indicate that it does well without upcasting. Therefore, I think we can just retain the original dtype in future commits.

Reply:
I was thinking about downcasting float64